Doc/reference/executionmodel.rst

```diff
@@ -163,14 +166,105 @@ These references follow the normal rules for name resolution with an exception
+but with some exceptions discussed below. :term:`Annotations <annotation>`
+currently do not use annotation scopes, but they are expected to use
+annotation scopes in Python 3.13 when :pep:`649` is implemented.
+
+Annotation scopes are used in the following contexts:
+
+* Type parameter lists for :ref:`generic type aliases <generic-type-aliases>`.
+* Type parameter lists for :ref:`generic functions <generic-functions>`. The function's annotations are
+  executed within the annotation scope, but its defaults and decorators are not.
+* Type parameter lists for :ref:`generic classes <generic-classes>`. The class's base classes and
+  keyword arguments are executed within the annotation scope, but its decorators are not.
+* The bounds and constraints for type variables
+  (:ref:`lazily evaluated <lazy-evaluation>`).
+* The value of type aliases (:ref:`lazily evaluated <lazy-evaluation>`).
+
+Annotation scopes differ from function scopes in the following ways:
+
+* If an annotation scope is immediately within a class scope, or within another
+  annotation scope that is immediately within a class scope, names defined in the
+  class scope can be accessed from within the annotation scope, as if the code in the
+  annotation scope was executing directly within the class scope. (By contrast, regular
+  functions defined within classes cannot access names defined in the class scope.)
+* Expressions in annotation scopes cannot contain :keyword:`yield`, ``yield from``,
+  :keyword:`await`, or :token:`:= <~python-grammar:expression>` expressions. (These
+  expressions are allowed in other scopes contained within the annotation scope.)
+* Names defined in annotation scopes cannot be rebound with :keyword:`nonlocal`
+  statements in inner scopes. This includes only type parameters, as no other
+  syntactic elements that can appear within annotation scopes can introduce new names.
+* While annotation scopes internally have a name, that name is not reflected in the
+  ``__qualname__`` of objects defined within the scope. Instead, the ``__qualname__``
+  of such objects is as if the object was defined in the enclosing scope.
+
+.. versionadded:: 3.12
+   Annotation scopes were introduced in Python 3.12 as part of :pep:`695`.
+
+.. _lazy-evaluation:
+
+Lazy evaluation
+---------------
+
+The values of type aliases created through the :keyword:`type` statement and
+the bounds and constraints of type variables created through the
+:ref:`type parameter syntax <type-params>` are *lazily evaluated*. This means
+that they are not evaluated when the type alias or type variable is created,
+but only when evaluation is necessary to resolve an attribute access.
+
+Example:
+
+.. doctest::
+
+   >>> type Alias = 1/0
+   >>> Alias.__value__
+   Traceback (most recent call last):
+     ...
+   ZeroDivisionError: division by zero
+
+Here the exception is raised only when the ``__value__`` attribute is accessed.
```

Comment:
Could we also have an example using `TypeVar` bounds?

Reply:
Adding